[wip] Expose DebugFission/Split-DWARF support in Frontend and Driver

weliveindetail opened the pull request:
This patch wires up flags in Driver and Frontend that expose support for [GNU debug fission](https://gcc.gnu.org/wiki/DebugFission) / [DWARF5 standardized Split-DWARF](http://www.dwarfstd.org/ShowIssue.php?issue=140421.1) as implemented in LLVM. It splits the majority of debug information from `.o` outputs into separate `.dwo` files.

weliveindetail committed:
[Frontend] Add Split-DWARF support (rev. 2)

I can not run the test yet, but this works:
```
S:\b\1\bin\swift-frontend.exe -frontend -c test.swift -o test.obj -g -split-dwarf-output test.dwo -sdk "S:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk"
```

weliveindetail committed:
[Driver] Add Split-DWARF support

weliveindetail committed:
[Frontend] Infer IRGenOptions.SplitDwarfOutput from FrontendOpts.InputsAndOutputs (and other fixes)

Other fixes:
* Add split-DWARF path to FrontendOptions::forAllOutputPaths()
* Fix test to use `-c` istead of `-emit-module -parse-as-library`
* Add test for multi-file invocation

weliveindetail committed:
[Frontend] Extend test DebugInfo/fission.swift

weliveindetail commented:
@swift-ci please test

compnerd reviewed (commented):
Could you also please add a test for the Jobs that are constructed?  I suspect that the use of this will require further plumbing.  The swift-driver is likely going to be more important for actual usage of this.

compnerd commented:
@swift-ci please test

weliveindetail committed:
[Driver] Fix target .dwo path with multiple outputs

chooseModuleAuxiliaryOutputFilePath() was setting ".dwo" due to ModulePath being empty.

weliveindetail committed:
[Driver] Assign DWARFFission from hasArg() directly

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

weliveindetail commented:
> Could you also please add a test for the Jobs that are constructed?

Do you mean I should extend the test(s) in `test/Driver/options.swift` or add a test somewhere else?

> I suspect that the use of this will require further plumbing. The swift-driver is likely going to be more important for actual usage of this.

Yes, adding support in `swift-driver` will need another PR in https://github.com/apple/swift-dirver

weliveindetail committed:
[Frontend] Write .dwo to temporary and store in target file only on finalize

weliveindetail commented on test/Driver/options.swift:
```diff
@@ -124,6 +124,13 @@
 // RUN: not %swiftc_driver -dwarf-version=6 %s 2>&1 | %FileCheck -check-prefix INVALID_DWARF_VERSION %s
 // INVALID_DWARF_VERSION: invalid value '{{1|6}}' in '-dwarf-version={{1|6}}'
 
+// RUN: %swift_driver -### -g -dwarf-fission -o %t/dwarf-fission.o %s 2>&1 | %FileCheck -check-prefix DWARF_FISSION %s
```
Looks like on Linux we invoke `bin/swift` here instead of `bin/swiftc`. Would I have to pass the `-Xfrontend` prefix before `-dwarf-fission` somehow or add support for it elsewhere? Or is this using https://github.com/apple/swift-dirver here?

The test failure was:
```
/home/build-user/swift/test/Driver/options.swift:128:19: error: DWARF_FISSION: expected string not found in input
// DWARF_FISSION: -split-dwarf-output {{.*}}/dwarf-fission.dwo
                  ^
<stdin>:1:1: note: scanning from here
<unknown>:0: error: option '-o' is not supported by 'swift'; did you mean to use 'swiftc'?
^
<stdin>:1:59: note: possible intended match here
<unknown>:0: error: option '-o' is not supported by 'swift'; did you mean to use 'swiftc'?
                                                          ^
```

Also, interesting that on Windows it's there:
```
> bin\swift --help | grep fission
  -dwarf-fission          Emit debug info as split-DWARF (Fission) in a separate .dwo file
```

hjyamauchi replied:
Not sure, but as I see this above at line 70, `%swift_driver` doesn't take `-o`? Do we need `%target-swift-frontend%` instead?

```
// RUN: not %swift_driver -### -o %t %s 2>&1 | %FileCheck -check-prefix=ARG_o %s
// ARG_o: error: option '-o' is not supported by 'swift'; did you mean to use 'swiftc'?
```

weliveindetail replied:
Haha, good point thanks! Did this in the evening in too much of a hurry. Let me revisit tomorrow.

weliveindetail committed:
[Frontend] clang-format (NFC)

weliveindetail committed:
[Driver] Fix tests in options.swift

weliveindetail committed:
[Frontend] Revert CodeGenFileType change (missing in 995e5a6)

compnerd commented:
@swift-ci please test

weliveindetail committed:
[Driver] Match temp file suffix in options.swift

weliveindetail committed:
[Driver] Match full path in options.swift

compnerd commented:
@swift-ci please test

weliveindetail commented:
Unrelated build error:
```
ModuleDependencies.cpp:406:24: error: no matching member function for call to 'emplace'
```

weliveindetail committed:
[Driver] Fix multi-unit test in options.swift

weliveindetail committed:
Merge branch 'main' into feature/dwarf-fission

weliveindetail commented:
Got this to build on my macBook and with the latest fix, tests are passing. @compnerd Could you give it one more try please?

weliveindetail commented:
I started to work on the argument processing in swift-driver, but haven't finished yet. It's pretty isolated and I don't think it will raise any questions that are relevant here.

weliveindetail commented:
The varying parameter/variable names might be confusing for reviewers. I am happy to change them if there are good reasons. Short notes on the current situation:

* The driver argument `-dwarf-fission` follows the recently added `-dwarf-version` argument. Previously, names for DWARF-specific arguments were prefixed with the debug `g` like `-gdwarf-types`. Clang uses `-gsplit-dwarf` and now has 2 values for it: `single` and `split`. If we implement single one day, it feels like this fits better with `-dwarf-fission`.

* The name of the frontend option name `OPT_emit_split_dwarf_path` follows the scheme for options that create supplementary outputs: https://github.com/apple/swift/blob/release/5.10/lib/Frontend/ArgsToFrontendOutputsConverter.cpp#L304

* The name of the frontend argument `-split-dwarf-output` is the name of the argument in both, llc and clang_cc1 (frontend). We are using the same code path in IRGen, so I thought it's good to keep the name aligned.

Maybe I should rename `SplitDwarfObjectPath` to `SplitDwarfOutputPath` in `SupplementaryOutputs`.

hjyamauchi commented:
@swift-ci please test

weliveindetail commented:
Development is going on in https://github.com/apple/swift/pull/71150
